### PR TITLE
Fix multiple JSON profiles issue in one model

### PR DESCRIPTION
**Pull request by aragentum**
Fixed support for several JSON profiles in one model.

Done:
- fixed bug #693
- renamed  `__profile__` to `__profiles__`
- added tests

If you have any comments or suggestions, let me know :)

### CODE REVIEW

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by aragentum** [comment]
@fantix I think we can't make the complexity any less.
Could you tell me how I can get into the authors of this project?

**Comment by wwwjfy** [comment]
Option 2: iterate every property of a model to find all JSON properties in `get_profile`.
Option 3: save all `prop_name` used during initialization and go through every one in `get_profile`.

The main difference is to keep `__profile__` flat.

As for author of the project, you are the author contributing MR. If you mean to become a maintainer, I think we don't have a rule or something now, but I believe it can be made by regularly answering questions and fixing issues for some time.

**Comment by aragentum** [comment]
@wwwjfy thank you for your suggestions and thank you for your work.

About your Option 2: I think it is not an optimal solution `get_profile` because this function called in many places and if we will search by all properties we can get performance problems then we will have a large number of objects/instances.

About your Option 3: Sorry, If I don't understand you correctly. We can add a separate place or we can add special property inside `__profile__` which will be store information about initialized `prop_names`. But I think both these solutions make the code more complex.

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by wwwjfy** [comment]
> About your Option 2: I think it is not an optimal solution because get_profile called in many places and if we will search by all properties we can get performance problems then we will have a large number of objects/instances.

To clarify, with the current implementation, the search is done once, and `__profile__` will be returned directly (well, it's also why we have this issue at the first place)

> About your Option 3: Sorry, If I don't understand you correctly. We can add a separate place or we can add special property inside __profile__ which will be store information about initialized prop_names. But I think both these solutions make the code more complex.

I mean when the table is initialized (in `_init_table` https://github.com/python-gino/gino/blob/c43b61956f2616011791bba87249c33c23091c07/src/gino/declarative.py#L286), we can know json properties and the corresponding columns. These info can be stored in a class attribute and here in `get_profile` we can use that information. 

At this moment I lean to option 3.

Both should be relatively small changes. I could have a diff to show my idea if that's more clear.

**Comment by aragentum** [comment]
@wwwjfy I get your point save to instance list of `prop_names` and in `get_profile` initialize all properties in one `__profile__`. Something like this (this is not final solution): 
```
sub_cls._prop_names = []
        rv = sa.Table(table_name, sub_cls.__metadata__, *args, **table_kw)
        for k, v in updates.items():
            setattr(sub_cls, k, v)
        for each_cls in sub_cls.__mro__[::-1]:
            for k, v in each_cls.__dict__.items():
                if isinstance(v, json_support.JSONProperty):
                    if v.prop_name not in sub_cls._prop_names:
                        sub_cls._prop_names.append(v.prop_name)
                    json_col = getattr(
                        sub_cls.__dict__.get(v.prop_name), "column", None
                    )
                    if not isinstance(json_col, sa.Column) or not isinstance(
                        json_col.type, sa.JSON
                    ):
                        raise AttributeError(
                            '{} "{}" requires a JSON[B] column "{}" '
                            "which is not found or has a wrong type.".format(
                                type(v).__name__, v.name, v.prop_name,
                            )
                        )
        return rv
```

**Comment by aragentum** [comment]
I will try to implement a new solution.

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
- src/gino/declarative.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
- src/gino/declarative.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by aragentum** [comment]
@wwwjfy what do you think about this solution? Codacy said that complexity has increased :(

Could you tell me `JSONProperties` will have aliases in the future? Because in this case properties in different profiles can have the same names (in my option this is not problem).

**Comment by fantix** [comment]
> Codacy said that complexity has increased :(

No worries, this is fine. :thisisfine:

**Comment by wwwjfy** [comment]
At least in major version 1, aliases are not a problem as the JSON properties are the same as "official" columns in the model classes. Same names can't happen.

Diff looks fine to me. @fantix please help to take a look too.

File: src/gino/declarative.py [hunk 1]
```diff
@@ -350,9 +350,13 @@ def _init_table(cls, sub_cls):
         rv = sa.Table(table_name, sub_cls.__metadata__, *args, **table_kw)
         for k, v in updates.items():
             setattr(sub_cls, k, v)
+
+        prop_names = []
```
Comment (fantix) [inline src/gino/declarative.py:354]: (minor) could use `set()` here

File: src/gino/declarative.py [hunk 2]
```diff
@@ -367,6 +371,7 @@ def _init_table(cls, sub_cls):
                                 type(v).__name__, v.name, v.prop_name,
                             )
                         )
+        sub_cls._prop_names = prop_names
```
Comment (fantix) [inline src/gino/declarative.py:374]: Let's use `__json_prop_names__` here

File: tests/test_json.py [hunk 3]
```diff
@@ -63,12 +69,14 @@ async def test_crud(bind):
     # Reload and test updating both JSON and regular property
     u = await User.get(u.id)
     await u.update(
-        age=User.age - 2, balance=200.15, realname="daisy", nickname="daisy.nick"
+        age=User.age - 2, balance=200.15, realname="daisy", nickname="daisy.nick", height=185, weight=75
```
Comment (fantix) [inline tests/test_json.py:72]: @daichen-daisy you're now 75kg and 185cm 🤣 just kidding, this is cool.

(trivial) Could you please also `black` this file? This line seems to be too long.

**Comment by aragentum** [comment]
@fantix sorry, are you mean remove last merge commit `Merge branch 'master' into bug/multiple-json-profiles`?

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
- src/gino/declarative.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
- src/gino/declarative.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
- src/gino/declarative.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/json_support.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5651113&bid=18502522)

**Comment by fantix** [comment]
Thanks for working on this one!

**Pull request merged**